### DISCUSSION TITLE
Promise.all to build new Firebase data

### DISCUSSION THREAD

**Discussion by GonzalesMatthew** [nss-evening-cohort-14/evening-client-side#216] [Look at what I learned!]
Dr. T helped me understand this better and asked I share with the class. I was making 'return' errors, and also encountering confusion on where to place my 'resolve' statement. Afterwards I was able to build new objects for posting based on data from two separate Firebase tables. I hope this helps anyone who is trying to do something similar.

First a bit of context. I was trying to create a master list of our Vendors and Rides which come from two separate FB tables. Why? Well because those tables had the names and prices for every vendor and ride at our park. If I could make a master list, then I could randomly assign one to one of our customers to simulate them buying food or riding a ride.

How would that work exactly? Well I would have to make a Promise.all call to get both vendors and rides and make my list.

Separately I would need to call my visitors and just get their identity. Then I could use both my list and my visitors to build new data for our 'Log' Firebase table which represents every transaction at our park. Makes sense, right?

To start, first I had to make my post function for my 'Log' data so that I can use it later.

```
// START SPENDINGMONEY(), WHICH CREATES LOGSARRAY AND POSTS TO FIREBASE
// ADD LOG
const addLog = (logArray) => new Promise(() => {
  logArray.forEach((entry) => {
    axios.post(`${dbUrl}/log.json`, entry);
  });
});
```
Now that that was out of the way I could focus on the fun part: making the master list. I used a Promise.all to get both of those tables. Then I destructured them with a loop to build a new array which I would then 'resolve' at the end so that I can use it later. That looks something like this.
```
// MASTER RIDES AND VENDORS LIST
const ridesAndVendors = () => new Promise((resolve, reject) => {
  Promise.all([getRides(), getVendors()])
    .then(([rides, vendors]) => {
      // BUILD ARRAY OF RIDES AND VENDORS TO CHOOSE FROM
      const ridesAndVendorsArray = [];
      rides.forEach((ride) => {
        const object = {
          earnings: ride.price,
          event: ride.rideName
        };
        ridesAndVendorsArray.push(object);
      });
      vendors.forEach((vendor) => {
        const object = {
          earnings: vendor.price,
          event: vendor.vendorName
        };
        ridesAndVendorsArray.push(object);
      });
      resolve(ridesAndVendorsArray);
    })
    .catch((error) => reject(error));
});
```
That wasn't so bad since we have done all of these things before, except maybe the Promise.all, which thankfully turns out to be fairly simple and straightforward.

Next I simply had to then use my newly created master list function inside another function so I could start to build our 'Log' data. In my new function, I called the master list function, then used a code block to also get my visitors data. Next, I wanted to be able to randomly choose one of my rides or vendors so that I could assign it to each visitor in a loop, so I made a randomizer function for my master list. Then I loop over my visitors and built a new object which also included a random ride or vendor. The tricky part you may see is the 'spread' operator which lets me quickly place my object inside another object as a key-value pair. See MDN docs for more info: 'Spread in object literals.' I then push that new object into an array to build an array of objects. Finally I ended my function by invoking my 'Log' post function I made at the beginning. If you read that function again you can see it accepts an array and iterates through each element, in this case objects,  and posts it to the 'Log' table.
```
// SPENDING MONEY
const spendingMoney = () => {
  ridesAndVendors().then((resultsArray) => {
    // (B) GET ALL VISITORS
    const logArray = [];
    getVisitors().then((visitorsArray) => {
    // START BUILDING LOG
      visitorsArray.forEach((visitor) => {
        const randomEvent = resultsArray[Math.floor(Math.random() * resultsArray.length)];
        const object = {
          visitor: visitor.visitorFirstName,
          ...randomEvent,
          timestamp: new Date()
        };
        logArray.push(object);
      });
      // (C) POST LOG TO FIREBASE
      addLog(logArray);
    });
  });
};
// END SPENDINGMONEY(), WHICH CREATES LOGSARRAY AND POSTS TO FIREBASE
```
Now I can assign spendingMoney() to a button click, and all of my guests will randomly choose an activity at the park. These activities get logged, which includes all the visitors, the events they picked individually and the associated cost. This helped us later where we needed to track our park's performance. Pretty slick, huh?

**Comment by drteresavasquez** [comment]
Excellent explanation!!
